### PR TITLE
Fixes #2: Case Class Schema

### LINKED ISSUE #2: Specialize caseClass1 - caseClassN
Currently, all the `Schema.caseClass` methods are written in terms of `Record`. This is semantically correct, but not optimally performant, because it involves packing and unpacking into maps.

To maximize performance of codecs generated with `Schema`, we should add `CaseClass` subtypes of `Schema` for various common arities of case classes (1 - 22???). Then as a convenience, we could have a method that converts these case classes to a `Schema.Record`, for cases where performance does not matter that much (e.g. `Schema.CaseClass2#toRecord`.

### PR DESCRIPTION
**Pull request by thinkharderdev**
@jdegoes Can you take a look at this guy and see if this is what you had in mind?

Also, for implementing the Json decoder, can we assume that JSON keys are ordered consistently with the case class declaration. AFAICT this  assumption is made by zio-json but not sure if that is a correct assumption.

### CODE REVIEW

**Comment by thinkharderdev** [comment]
JsonCodec is implemented. Still need to do the ProtobufCodec.

File: core/src/main/scala/zio/schema/Schema.scala
Comment (jdegoes) [inline core/src/main/scala/zio/schema/Schema.scala:52]: I think this encoding is almost perfect. After studying it for a while, I can see that `extract` will necessarily allocate a tuple. We can eliminate this overhead by switching to a slightly different encoding:

```scala
 case class CaseClass3[A1, A2, A3, Z](
     field1: (String, Schema[A1]),
     field2: (String, Schema[A2]),
     field3: (String, Schema[A3]),
     construct: (A1, A2, A3) => Z,
     extractField1: Z => A1,
     extractField2: Z => A2,
     extractField3: Z => A3
   ) extends CaseClass[Z] 
```

Then we could pick apart the `Z` into its component pieces without doing any allocations; and we can already, having all the component pieces, construct the `Z` by using `construct` without any allocations.

File: core/src/main/scala/zio/schema/Schema.scala
Comment (jdegoes) [inline core/src/main/scala/zio/schema/Schema.scala:46]: Here and elsewhere, I'd use:

```scala
final case class
```

Since it does not make sense to override case classes.

File: core/src/main/scala/zio/schema/Schema.scala
Comment (jdegoes) [inline core/src/main/scala/zio/schema/Schema.scala:429]: I love the `toRecord` on all of these, because it gives library authors a convenient way to generically handle all of them, albeit with performance penalty.

File: core/src/main/scala/zio/schema/Schema.scala
Comment (jdegoes) [inline core/src/main/scala/zio/schema/Schema.scala:30]: Making sure all the case class types extend this is perfect! This will allow more uniform, type safe handling in user-land code.

File: core/src/main/scala/zio/schema/codec/ProtobufCodec.scala
Comment (jdegoes) [inline core/src/main/scala/zio/schema/codec/ProtobufCodec.scala:387]: Quick fix here and elsewhere will be `case caseClass : CaseClass[A] => decoder(caseClass.toRecord)`

**Comment by jdegoes** [comment]
@thinkharderdev This is great! A few minor comments but I am ready to merge when you're happy with it.

File: core/src/main/scala/zio/schema/codec/ProtobufCodec.scala
Comment (thinkharderdev) [reply core/src/main/scala/zio/schema/codec/ProtobufCodec.scala]: This won't work because you end up with a `Decoder[Map[String,_]]` instead of a `Decoder[A]`

I think it shouldn't be too hard to implement the protobuf encoder/decoder for the case class schemas though.

File: core/src/main/scala/zio/schema/Schema.scala
Comment (thinkharderdev) [reply core/src/main/scala/zio/schema/Schema.scala:55]: I like this a lot. Should set us up well to implement Optics as well.

**Comment by thinkharderdev** [comment]
@jdegoes Addressed review comments. Protobuf encoder/decoder is still unimplemented though. I can knock that out before we merge or I can do a follow-up PR if you'd rather get this in main

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
@thinkharderdev I will merge now to avoid conflicts (it's a pretty big PR). Thanks for your work on this, and look forward to the follow-up PR!

**Pull request merged**
